Adrian AI Materials (#5)
Ada-Developers-Academy/independent-ai-topics: working-with-openai/setting-up-open-ai-postman.md

## The OpenAI API and Postman

As a reminder, when we work with APIs, Postman is a great tool for testing our endpoints, making sure we are formatting our requests correctly, and ensuring that we are receiving the expected responses. 

In the past, when we worked with APIs that require authorization, we created an environment within our project using the .env file where we stored our environment variables such as our API key. To make testing out OpenAI endpoints in Postman just a bit easier, we can create an environment within Postman itself that includes our API key so that we don't have to manually attach it to each request. We can do that following the steps below.

## Setting up OpenAI in Postman

1. Open up an untitled POST request. Click the drop down for "No environment" and select "New Environment". Nothing will happen except what gets selected in the drop down.  
![New Environment Drop Down in Postman](assets/setting-up-openai-postman-and-java/open-ai-postman-new-environment.png)  
*Fig. Postman POST request with New Environment Selected*

2. Set the URL in Postman to the following endpoint: `https://api.openai.com/v1/chat/completions`.   

> We will talk more in the next lesson about what this request is and how it works, but for now, we will use it to verify that our environment is set up correctly. If you try and make the request now, you will get the following response body with an error that states you are missing an API key.  

Reviewer: ```suggestion
We will talk more in a future lesson about what this request is and how it works, but for now, we will use it to verify that our environment is set up correctly. If you try and make the request now, you will get the following response body with an error that states you are missing an API key.  
```

> # Setting up the OpenAI API in Postman

Reviewer: It might make it a little easier for students to find information if we differentiate the lesson names, especially since longer names can be truncated in the side bar. A couple potential ideas:

```suggestion
# Calling the OpenAI API from Postman
```

```suggestion
# Sending requests to the OpenAI API from Postman
```

## Goals
Now that we have our OpenAI account created and verified, let's get OpenAI set up in Postman and Python so we are prepared to start testing our API calls and bringing them into our project.

Our goals for this lesson are to:
- Set up an OpenAI environment in Postman
- Make our first Postman API 

> - Learn how to set up the OpenAI library in Python

Reviewer: I missed before that we don't do the python set up here, so we should remove that goal.

```suggestion
Our goals for this lesson are to:
- Set up an OpenAI environment in Postman
- Make our first request to the OpenAI API 
```

## The OpenAI API and Postman

> As a reminder, when we work with APIs, Postman is a great tool to test our endpoints, make sure we are formatting our requests correctly and ensure that we are receiving the expected responses. 

Reviewer: ```suggestion
As a reminder, when we work with APIs, Postman is a great tool for testing our endpoints, making sure we are formatting our requests correctly, and ensuring that we are receiving the expected responses. 
```

In the past, when we worked with APIs that require authorization, we created an environment within our project using the .env file where we stored our environment variables such as our API key. To make testing out OpenAI endpoints in Postman just a bit easier, we can create an environment within Postman itself that includes our API key so that we don't have to manually attach it to each request. We can do that following the steps below.

## Setting up OpenAI in Postman

1. Open up an untitled POST request. Click the drop down for "No environment" and select "New Environment". Nothing will happen except what gets selected in the drop down.  
![New Environment Drop Down in Postman](assets/setting-up-openai-postman-and-java/open-ai-postman-new-environment.png)  
*Fig. Postman POST request with New Environment Selected*

> 2. Add the following request to Postman: https://api.openai.com/v1/chat/completions.   

Reviewer: ```suggestion
2. Set the URL in Postman to the following endpoint: `https://api.openai.com/v1/chat/completions`.   
```

![Open AI Postman Model Parameters Needed Error](assets/setting-up-openai-postman-and-java/open-ai-postman-model-parameters-needed-error.png)  
*Fig. Model parameters needed error*  

If you get this error, it's a pretty sure sign your API key is working. However, let's add a request body to make sure we are able to get a working response.  Go ahead and navigate to the body tab and select "raw" with JSON as the format. Copy and paste the following request body into the window:

```py
{
    "model" : "gpt-3.5-turbo",
    "messages" : [
        {
            "role" : "user",
            "content" : "Write me a list of 5 flightless birds."
        }
    ]
}
```
You should now be able to submit your POST request and get a response more in line with what you are looking for. Your response may look slightly different, but a sample of what you could see is shown below:  

![Successful API response](assets/setting-up-openai-postman-and-java/open-ai-postman-response-success.png) 
*Fig. Successful OpenAI POST request with response body ([Full size image](assets/setting-up-openai-postman-and-java/open-ai-postman-response-success.png))*  

> If you've made it this far, congratulations! You now have Postman configured to practice the openAI requests that we will be going over in the next couple of lessons. If you saved your environment correctly, it should now exist as an option in Postman when you choose your environment. If for any reason your requests still aren't working, double check your work or reach out to an instructor for further assistance! 

Reviewer: To encourage folks to reach out to the wider community vs directly to instructors, we could tweak the last sentence.

```suggestion
If you've made it this far, congratulations! You now have Postman configured to practice the openAI requests that we will be going over in the next couple of lessons. If you saved your environment correctly, it should now exist as an option in Postman when you choose your environment. If for any reason your requests still aren't working, double check your work then use resources such as #study-hall, office hours, etc. for further assistance! 
```

### !callout-warn

## API Calls in Postman Use Credits

> Just a quick reminder that any API calls you make to OpenAI will use credits. You can check your usage in your account online.

Reviewer: This isn't necessary, but I feel like it couldn't hurt to give a little more explicit reminder of where to check since they've only seen where to check their credits in the curriculum once.

```suggestion
Just a quick reminder that any API calls you make to OpenAI will use credits. You can check your credit balance on the "Usage" page of your account online.
```
